src/julia_internal.h

```diff
@@ -842,12 +843,13 @@ jl_method_t *jl_make_opaque_closure_method(jl_module_t *module, jl_value_t *name
     int nargs, jl_value_t *functionloc, jl_code_info_t *ci, int isva, int isinferred);
 JL_DLLEXPORT int jl_is_valid_oc_argtype(jl_tupletype_t *argt, jl_method_t *source);
 
-STATIC_INLINE int is_anonfn_typename(char *name)
+STATIC_INLINE int is_anonfn_typename(char *name) JL_NOTSAFEPOINT
 {
     if (name[0] != '#' || name[1] == '#')
         return 0;
     char *other = strrchr(name, '#');
-    return other > &name[1] && other[1] > '0' && other[1] <= '9';
+    // Check if the last character is a digit
+    return other > &name[1] && isdigit(name[strlen(name) - 1]);
```

Comment:
`#bar##foo2` doesn't seem possible, since it would happen if the outermost function were named `#foo` or alike. 

Note that `fl_current_module_counter` explicitly avoids inserting the outermost function name into the mangled name if the outermost function name starts with a hash.

Reply:
Got it - What was the example that motivated the change in that case?

Reply:
Yeah, this PR relies on `##<digit>+` not appearing somewhere else in the name mangling.

Reply:
I mean what's the case where 0-9 no longer follows the last `#` in the anon fn name?

Reply:
Ah, OK. Yes, we don't need to refactor this function anymore. The change is likely reminiscent from the version that was using `<>` as the separator.

Reply:
Yes, just making the last line `return other > &name[1] && isdigit(other[1]);` should probably work (`isdigit` is needed since we may get something like `#bar#foo##0`).